pipeline/pipeline/pipeline.py

```diff
@@ -82,19 +135,107 @@ def _new_resource_group(self, source, mappings):
         return rg
 
     def read_input(self, path, extension=None):
+        """
+        Create a new input resource file object representing a single file.
+
+        Examples
+        --------
+
+        Read the file `hello.txt`:
+
+        >>> p = Pipeline()
+        >>> input = p.read_input('hello.txt')
+        >>> t = p.new_task()
+        >>> t.command(f"cat {input}")
+        >>> p.run()
+
+        Parameters
+        ----------
+        path: :obj:`str`
+            File path to read.
+        extension: :obj:`str`, optional
+            File extension to use.
+
+        Returns
+        -------
+        :class:`.InputResourceFile`
+        """
+
         irf = self._new_input_resource_file(path)
         if extension is not None:
             irf.add_extension(extension)
         return irf
 
     def read_input_group(self, **kwargs):
+        """
+        Create a new resource group representing a mapping of identifier to
+        input resource files.
+
+        Examples
+        --------
+
+        Read a binary PLINK file:
+
+        >>> p = Pipeline()
+        >>> bfile = p.read_input_group(bed="data/example.bed",
+        ...                            bim="data/example.bim",
+        ...                            fam="data/example.fam")
+        >>> t = p.new_task()
+        >>> t.command(f"plink --bfile {bfile} --geno --out {t.geno}")
+        >>> t.command(f"wc -l {bfile.fam}")
+        >>> t.command(f"wc -l {bfile.bim}")
+        >>> p.run()
+
+        Parameters
+        ----------
+        kwargs: :obj:`dict` of :obj:`str` to :obj:`str`
+            Key word arguments where the name/key is the identifier and the value
+            is the file path.
+
+        Returns
+        -------
+        :class:`.InputResourceFile`
+        """
+
         root = self._tmp_file()
         new_resources = {name: self._new_input_resource_file(file, root + '.' + name) for name, file in kwargs.items()}
         rg = ResourceGroup(None, root, **new_resources)
         self._resource_map.update({rg._uid: rg})
         return rg
 
     def write_output(self, resource, dest):  # pylint: disable=R0201
+        """
+        Write resource file or resource file group to an output destination.
+
+        Examples
+        --------
+
+        Write a single task intermediate to a permanent location:
+
+        >>> p = Pipeline()
+        >>> t = p.new_task()
+        >>> t.command(f'echo "hello" > {t.ofile}')
```

Comment:
I presume there's nothing special about `.ofile` here? Thinking from a user perspective, may want to make that clear somewhere.

Reply:
Not sure what you mean by nothing special. It's implicitly creating a TaskResourceFile that represents a temporary file. This can then be referred to with `t.ofile` in the same context as when you'd normally pass a file path.

```
t.command(f'echo "hello" > {t.tmp1}') # attribute tmp1 on t doesn't exist and is implicitly created
t.command(f'cat {t.tmp1} > {t.ofile}') # attribute tmp1 on t now exists
```

Reply:
Yep, that's exactly my point. It doesn't have to be `.ofile` but could be `.ofile5000` (user-specifiable, not a fixed field)

Reply:
I... meant anything random - not sure `.ofile5000` was actually the best example 😛